src/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -355,60 +370,108 @@ public static boolean isValidFile(final InputStream stream)
+            if (headerByteCount == 0) {
+                // Handle case where there is no empty gzip block at end.
+                return new DecompressedBlock(blockAddress, new byte[0], 0);
+            }
+            if (headerByteCount != BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH) {
+                return new DecompressedBlock(blockAddress, headerByteCount, new IOException("Premature end of file"));
+            }
+            final int blockLength = unpackInt16(mFileBuffer, BlockCompressedStreamConstants.BLOCK_LENGTH_OFFSET) + 1;
+            if (blockLength < BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH || blockLength > mFileBuffer.length) {
+                return new DecompressedBlock(blockAddress, blockLength,
+                        new IOException("Unexpected compressed block length: " + blockLength));
+            }
+            final int remaining = blockLength - BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH;
+            final int dataByteCount = readBytes(mFileBuffer, BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH,
+                    remaining);
+            mStreamOffset += dataByteCount;
+            if (dataByteCount != remaining) {
+                return new DecompressedBlock(blockAddress, blockLength,
+                        new FileTruncatedException("Premature end of file"));
+            }
+            final byte[] decompressed = inflateBlock(mFileBuffer, blockLength, bufferAvailableForReuse);
+            return new DecompressedBlock(blockAddress, decompressed, blockLength);
+        } catch (IOException e) {
+            return new DecompressedBlock(blockAddress, 0, e);
         }
-        inflateBlock(mFileBuffer, blockLength);
-        mCurrentOffset = 0;
-        mBlockAddress += mLastBlockLength;
-        mLastBlockLength = blockLength;
     }
 
-    private void inflateBlock(final byte[] compressedBlock, final int compressedLength)
-        throws IOException {
-        final int uncompressedLength = unpackInt32(compressedBlock, compressedLength-4);
-        byte[] buffer = mCurrentBlock;
-        mCurrentBlock = null;
-        if (buffer == null || buffer.length != uncompressedLength) {
-            try {
-                buffer = new byte[uncompressedLength];
-            } catch (final NegativeArraySizeException e) {
-                throw new RuntimeIOException("BGZF file has invalid uncompressedLength: " + uncompressedLength, e);
-            }
+    private byte[] inflateBlock(final byte[] compressedBlock, final int compressedLength,
+            final byte[] bufferAvailableForReuse) throws IOException {
+        final int uncompressedLength = unpackInt32(compressedBlock, compressedLength - 4);
+        if (uncompressedLength < 0) {
+            throw new RuntimeIOException("BGZF file has invalid uncompressedLength: " + uncompressedLength);
+        }
+        byte[] buffer = bufferAvailableForReuse;
+        if (buffer == null || uncompressedLength != buffer.length) {
```

Comment:
add a comment to the effect that we must allocate a new buffer because the one that is available is no good